Feature sell rabbits

HenrikeW opened the pull request:
A part of the form used for adding new animals was turned into a reusable component and now even serves for selling individuals. 
The button for selling individuals in `individual_view` should only appear if you have the permission to sell this animal (i.e. if you are an admin, the genebank manager or the owner of the individual's current herd). 
There is a bug related to this described in issue #334 . 

I tested the feature manually with an admin user and it works fine. 
When I tried with a user who only owns a certain herd, I got a 502 response back and can't find the cause of that problem. 
I'd be happy if you could give it a try as well.

HenrikeW committed:
move selling form to its own component

copy JSX and everything necessary from individual_add.
add props, adjust styling.

HenrikeW committed:
replace JSX in individual_add with selling form component

also fixing minor bug in createBreeding.

HenrikeW committed:
create sell-individual component and import it

HenrikeW committed:
add information text, handleUpdateIndividual and clear herd input

HenrikeW committed:
add patch request and sell button

HenrikeW committed:
fix undefined autocomplete in selling form

HenrikeW committed:
styling, ux and error handling for selling rabbits

HenrikeW committed:
only send necessary fields to api/individual

instead of the whole individual object, only the fields number, herd and selling date are sent.

pontus commented:
Would it be possible to remove/disable the sell button once there is a proper submitted sale? Once the sale goes through, a herd owner no longer has access to the animal and might be confused and try to sell it again.

pontus commented on frontend/src/individual_sellingform.tsx:
```diff
@@ -0,0 +1,76 @@
+  },
+});
+
+export function IndividualSellingform({
```
```suggestion
export function IndividualSellingForm({
```

HenrikeW replied:
This change makes sense, of course. Renamed it.

pontus commented on frontend/src/individual_add.tsx:
```diff
@@ -28,6 +28,7 @@ import { useMessageContext } from "@app/message_context";
 import { get, post } from "@app/communication";
 import { useUserContext } from "./user_context";
 import { useDataContext } from "./data_context";
+import { IndividualSellingform } from "./individual_sellingform";
```
```suggestion
import { IndividualSellingForm } from "./individual_sellingform";
```

pontus commented on frontend/src/individual_add.tsx:
```diff
@@ -493,45 +491,12 @@ export function IndividualAdd({
-                onChange={(event, newValue) => {
-                  newValue && handleUpdateIndividual("herd", newValue.herd);
-                }}
+              <IndividualSellingform
```
```suggestion
              <IndividualSellingForm
```

pontus commented on frontend/src/individual_sellingform.tsx:
```diff
@@ -0,0 +1,76 @@
+
+export function IndividualSellingform({
+  individual,
+  herdOptions,
+  herdKey,
```
I'm not thrilled with these names. And what is `herdKey` here?

HenrikeW replied:
`herdOptions` is a way of naming that has been present in the code base since I started working with it, there are `colorOptions`, `sexOptions` and maybe some more in similar use cases, so I'd rather not change it here. 

The `herdKey` is used to reset the Autocomplete from Material UI when the selling-form-component is rendered in `individual_add`. To clear the Autocomplete when the user clicks on "Ny kanin", I change the key to cause a rerender of the component. It was the only  way I managed to clear the Autocomplete and make it show the placeholder text ("Välj besättning") and not just "undefined". Found here: https://stackoverflow.com/questions/59790956/material-ui-autocomplete-clear-value Maybe there is a better way, but I tried a lot. 
I'm open to a name it differently but thought this was most appropriate. What would you suggest?

pontus replied:
Ok, good enough.

HenrikeW committed:
rename IndividualSellingform into IndividualSellingForm

HenrikeW commented:
> Would it be possible to remove/disable the sell button once there is a proper submitted sale? Once the sale goes through, a herd owner no longer has access to the animal and might be confused and try to sell it again.

The button should only be visible if the individual you look at belongs to your herd, i.e. if you actually can sell it. Once you sold it, the button should disappear. 
If you still see the button, it might be related to a bug I discovered while building this. I described it here https://app.zenhub.com/workspaces/aspebodaklungan-60507bca37b5030019f0c5a4/issues/nbisweden/theherdbook/334 but haven't fixed it yet.

pontus commented:
I may have been unclear and don't know if we're talking about the same button here - when selling an animal, the form wasn't automatically closed for me, and the "Sälj" button was still accessible in there (and I don't see anything that I understand would unrender it).

I never checked (at least not so I remember it ) on the individual page, but that is wrapped in a `canEdit` so I believe that is fine.

But for clarity; it's the button within the popup I'm referring to.

HenrikeW committed:
add success box instead of user message after sale

THis way, the button "Sälj" disappears and the user can't make any invalid requests once the individual has been sold.

HenrikeW commented:
In that case I misunderstood you. But good point! It makes sense to let that button disappear. I know replaced the user message that popped up in the bottom of the page with a success box that appears within the popup.
